eth/backend.go

```diff
@@ -538,46 +568,51 @@ func (s *Ethereum) Protocols() []p2p.Protocol {
 // Start implements node.Lifecycle, starting all internal goroutines needed by the
 // Ethereum protocol implementation.
 func (s *Ethereum) Start() error {
-	eth.StartENRUpdater(s.blockchain, s.p2pServer.LocalNode())
-
+	if !s.readOnly {
+		eth.StartENRUpdater(s.blockchain, s.p2pServer.LocalNode())
+	}
 	// Start the bloom bits servicing goroutines
 	s.startBloomHandlers(params.BloomBitsBlocks)
 
 	// Regularly update shutdown marker
 	s.shutdownTracker.Start()
 
-	// Figure out a max peers count based on the server limits
-	maxPeers := s.p2pServer.MaxPeers
-	if s.config.LightServ > 0 {
-		if s.config.LightPeers >= s.p2pServer.MaxPeers {
-			return fmt.Errorf("invalid peer config: light peer count (%d) >= total peer count (%d)", s.config.LightPeers, s.p2pServer.MaxPeers)
+	if s.readOnly {
+		// Figure out a max peers count based on the server limits
+		maxPeers := s.p2pServer.MaxPeers
+		if s.config.LightServ > 0 {
+			if s.config.LightPeers >= s.p2pServer.MaxPeers {
+				return fmt.Errorf("invalid peer config: light peer count (%d) >= total peer count (%d)", s.config.LightPeers, s.p2pServer.MaxPeers)
+			}
+			maxPeers -= s.config.LightPeers
 		}
-		maxPeers -= s.config.LightPeers
+		// Start the networking layer and the light server if requested
+		s.handler.Start(maxPeers)
 	}
-	// Start the networking layer and the light server if requested
-	s.handler.Start(maxPeers)
 	return nil
 }
 
 // Stop implements node.Lifecycle, terminating all internal goroutines used by the
 // Ethereum protocol.
 func (s *Ethereum) Stop() error {
-	// Stop all the peer-related stuff first.
-	s.ethDialCandidates.Close()
-	s.snapDialCandidates.Close()
-	s.handler.Stop()
-
+	// Only Stop P2P when P2P was initiated
+	if !s.readOnly {
+		// Stop all the peer-related stuff first.
+		s.ethDialCandidates.Close()
+		s.snapDialCandidates.Close()
+		s.handler.Stop()
+	}
```

Comment:
I don't know if this is easy to code, but it'd be nice not to branch on `!s.readOnly` in particular, but rather have the `Close()`/`Stop()` functions handle a never-opened/never-started case more gracefully.

Analogous to how the destructor of a unique_ptr just does the right thing even if the unique_ptr is null.